Peer discovery for compact block filters

johncantrell97 opened the pull request:
### Description

This adds a new p2p discovery module to the compact block filters blockchain implementation in BDK.  The high level idea is for a wallet that wants to be a light client using compact block filters to not need to specify an explicit peer to sync the chain.

The current implementation starts with the DNS seeds and performs a network search using `GetAddr` messages with connected peers.  It builds a directory with successfully connected addresses and their services flag.  It caches this directory to the filesystem to speed up discovery on subsequent startups of the wallet.

Currently there is no way to extract a peer or set of peers from this directory but that is a good next step.

It currently uses a configurable amount of worker threads to perform the crawl and I recommend spawning a thread to run the `manage` control loop that spawns the workers and handles worker messages.

The high level structs and architecture looks like this:

`AddressManager` -- the top level struct that gets instantiated to start everything.  The manager holds the directory, a shared list of addresses being searched, and channels for messaging between the worker threads and the manager.

`AddressDirectory` -- Right now it holds two indexes, one for cbf enabled addresses and the other for the rest.  Each index is a `VecDeque<NodeAddress>` where a `NodeAddress` is just an SocketAddr and ServicesFlag.  The indexes probably should not be VecDeque's, this was just easier for serializing from this to the filesystem and back into a the same structure the workers use for managing a list of addresses to search.  

`AddressDiscovery` -- a struct wrapped in Arc<Mutex> for sharing between worker threads.  It contains a list of addresses to be searched and which addresses have already been visited.  

`AddressWorker` -- a struct representing a network crawler worker.  it maintains a reference to the AddressDiscovery struct and a channel for sending messages to the AddressManager.

`DiscoveryProgress` -- a trait for receiving progress updates from the address manager.  this is modeled after the Progress trait used to track progress when syncing the blockchain. I provided implementations for noop, logger, and Sender<DiscoveryData>.

The changes to `peer` were to prevent the workers from getting stuck on a particular peer.  The current implementation blocks indefinitely when connecting and when receiving responses.  I added a connect_with_timeout and also changed the verack and version messages to timeout after 1 second.  We probably don't want to keep it exactly like this.  It might not even make sense to use the Peer struct when doing the discovery.  It also might make sense to rewrite some of the peer network messaging logic in a way that doesn't block and expect a single response.

Some example code if you want to try running an AddressManager search with BDK from this fork:

```
use std::thread;
use bdk::blockchain::compact_filters::{AddressManager, log_discovery_progress};
use bdk::bitcoin::network::constants::Network;
use simple_logger::SimpleLogger;

fn main() -> Result<(), bdk::Error> {
    SimpleLogger::new().init().unwrap();
    
    let network = Network::Bitcoin;
    let mut address_manager = AddressManager::new(network, "cache.json".to_string(), 8);
    
    let manager_handle = thread::spawn(move || {
        address_manager.manage(log_discovery_progress());
    });

    manager_handle.join().unwrap();
    Ok(())
}
```


### Notes to the reviewers

This is not really ready for a real review.  This is to help start documenting the idea and get Raj (+others?) in the loop as early as possible.

### Checklists

#### All Submissions:

* [ ] I've signed all my commits
* [ ] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [ ] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

rajarshimaitra commented:
@johncantrell97 thank for working on this. This created a nice base to introduce the rest of the cbf functionalities.

I have been trying over the p2p manager, and have finalized how the address manager would look like. Added some changes, and updated via logically occurring commit. Below is a description of the change and rationales.

### Objective ###
In my mind a `PeerManager` should have a peer directory and an `AddressManager` (this PR). The sole purpose of `AddresssManager` is to handle crawling through the network and fetch valid and working node addresses. 

The Major interface between `AddressManager` and the user is via a `fetch()` method, which takes in number of cbf/non-cbf peers and provide freshly crawled addresses to user. 

Address manager should maintain a cache (file storage, json for now) and should effectively use it to quickly bootstrapped from previously found nodes. Caching should be done at the end of every fetch, and cache should only update if previously unknown address is found.

Address manager should not blindly trust previously found nodes, and should connect to verify they are still alive. For this reason `AddressDirectory` is the live in-memory store of freshly found addresses, and each instance of a manager will start with fresh `AddressDirectory`.

Once it has a populated cache, subsequent fetches should be fast. This is done by converting the existing cache into seeds and putting the previously known cbf addresses at front, so they get tried out first before non-cbf addresses. As the major time is taken for finding cbf peers in the network, this speeds up the process.  

Given the above objective and after few trials below is the `AddressManager` architecture I ended up settling for. Which is mostly similar to @johncantrell97 's initial sketch with some modifications.

![image](https://user-images.githubusercontent.com/36541669/122651139-6f517700-d154-11eb-9fe1-f248c81fc950.png)

### Change synopsis ###
Commits are atomic and should be easy to review them independently.

Added documentation.

Removed public visibility of most structures, as they will only be exposed to the manager.

Added a test case to demonstrate how the manager should be used to fetch fresh addresses from the network.

Each time `fetch()` is called, the manager will spawn new workers who will populate the directory, starting with a seed create from already existing cache.

The test includes a timer to compare time taken fetch 5 cbf and 5 non-cbf peers, once from and empty cache, and then from the populated cache of the first run.

In my local it took
- first crawl : 125 secs
- second crawl - 2.5 secs

Once the manager finds some cbf peers and cache them, subsequent fetches are mostly fast.

rajarshimaitra commented:
I am getting this error, randomly for some peers
```
<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })', src/blockchain/compact_filters/p2p.rs:185:38
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
   3: core::result::Result<T,E>::unwrap
             at /home/raj/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1037:23
   4: bdk::blockchain::compact_filters::p2p::AddressWorker::work
             at ./src/blockchain/compact_filters/p2p.rs:185:25
   5: bdk::blockchain::compact_filters::p2p::AddressManager<P>::spawn_workers::{{closure}}
             at ./src/blockchain/compact_filters/p2p.rs:428:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

It means `peer.close()` is being called on a peer which is not connected. Not sure why that would be the case. Will dig deeper. Suggestions are most welcome.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+                    if let Ok(peer) = potential_peer {
+                        peer.send(NetworkMessage::GetAddr).unwrap();
+                        self.try_receive_addr(&peer);
+                        self.try_receive_addr(&peer);
```
Are you supposed to receive two `addr` messages or is this just to give it one more chance to pickup the message before giving up?

rajarshimaitra replied:
I guess it was to give it one more chance. I removed it in the last changes. But I am not fully sure that it doesn't have any impact.

I am having trouble syncing from the existing seed. New addresses are not updating in the pending list. But it works with a previously populated cache.

It could be related to this. Will investigate more.

johncantrell97 replied:
A peer can respond to get addresses in multiple replies and the current peer code ever only expects one reply.  We should probably at some point refactor some of the peer communication layer to be a bit more dynamic but until then I thought it might be useful to poll twice for addr replies.  

I think I logged it at some point and did see it receiving addresses on the second try so it might be helpful.

rajarshimaitra replied:
Yes makes sense. At one point I have seen (without the second try_receive) that new pendings are not being updated in the discovery. Or it's getting updated with just 1 new address. As a result, the discovery pending list is being quickly exhausted and the workers get stuck.

I changed it back to double try and it didn't happen after that. But in any case, we should handle that situation properly. What to do when the pending list depletes in the discovery and workers don't have any more addresses to try. This can happen for multiple other reasons too, ex: A worker got new addresses and waiting for the write lock. In the meantime, other workers deplete the pending list and then nobody has any new addresses to try and they all wait.

To mitigate part of this I changed the pending addition task from the manager to worker. So instead of one manager thread trying to add pending to discovery it's all the worker threads and hopefully, someone will get to add new pending before the list depletes.

It's still nasty if that happens. There is no run time error, everything just pauses, and the user doesn't have any clue what went wrong.

The above refactoring removed the need for `AddressMessages` as it's just only one type of message the worker will send to the manager.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+            Err(_) => return None,
+        };
+
+        let address_cache = serde_json::from_str(&serialized).unwrap();
```
I would avoid the `unwrap()` here, either change the fn to return a `Result` or just return `None` like you do for the read errors

rajarshimaitra replied:
I am replacing unwraps from all the places, and creating error structure for `Peer`, `AddressManager`, `PeerManager` etc and then combined them into `CompactFiltersErrror`. So will handle all the unwraps in that following work.

We should have proper unwrap handling everywhere.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+    fn write_to_file(&self, path: &str) -> Result<(), std::io::Error> {
+        let serialized = serde_json::to_string_pretty(&self).unwrap();
+
+        let mut cache_file = File::create(path).unwrap();
```
Same here, in a fn that returns a result there shouldn't really be any `unwraps`

rajarshimaitra replied:
https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660346200

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+#[derive(Serialize, Deserialize)]
+struct AddressCache {
+    cbf: Vec<SocketAddr>,
+    non_cbf: Vec<SocketAddr>,
```
I would change these to `HashSets`, I don't think the ordering of the items it's particularly important. This would have the advantage that loading the same cache multiple times effectively shuffles the items...

rajarshimaitra replied:
Yup cache is now updated to `HashSet`.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+            .iter()
+            .chain(self.non_cbf.iter())
+            .copied()
+            .collect()
```
... to avoid using the same few seeds over and over again

rajarshimaitra replied:
I am not sure I understand. 

What we are doing here is ensuring the CBF addresses are in front of the seed vector. In that way, those are the addresses the worker will try first, even before existing seeds.

Because finding cbf peers are much slower than finding regular peers, the discovery speeds up if we try already known cbf addresses first.

afilini replied:
I was thinking that if the peers are stored in `Vec`s every call to this method would return the same few peers in the same order, unless they are explicitly removed from the cache.

If the caller uses a logic like "as soon as I reach 5 valid peers I'll stop looking" returning the seeds in the same order could lead to the caller just re-using the same peers over and over again, even across shutdowns.

Using `HashSet`s somewhat mitigates this I think.

rajarshimaitra replied:
Ya, once we have `HashSet` the ordering is jumbled up. 

But it's still true that it's the same peers from the previous fetch that were cached in the last run.

I think we cannot do away with that repetition (that's the point of a cache anyway), So the manager checks the connection again, see if the peer is alive, see if it's in the ban list, check weather cbf/non_cbf, and then pass it on to the caller.

So at the end of all of it, we can be sure that the manager didn't just hand over a dead peer just because it had it in the cache last time.

rajarshimaitra commented:
Made a lot of updates to both API and internal structures. The objective is to make address crawling faster.

- Rebased the PR so previous commits also updated. Which doesn't need re-review. Previous iteration's last commit was https://github.com/bitcoindevkit/bdk/pull/378/commits/bfc13587bf1cfc1bc1854d962ea7027a3cce089e
So everything after that is new additions.

- Manager is modified to maintain a specific reserve of fresh addresses. (For now, the default is 5 for both cbf and non_cbf). The directory will keep track of all the previously supplied addresses to the user and will start workers to replenish the fresh buffer.

- The major API of the address manager is `fetch()`, which when called will fetch address from the network until buffer requirement is reached.

- Decided not to go with its own thread. So fetch will block the user thread until it returns. 

- Added banning logic to the manager and other internal structures.

- Major structural changes to make various comparisons and insertion efficient. So far I think I am quite satisfied with the speed of the crawl. the timer tests now return 35 secs and 1.5 secs for fresh and cached fetch respectively. Which I think is satisfactory. There are probably more possible optimization exists. Open for suggestions.

So far I think it finally has the right APIs to work with a PeerManager. Needs much housekeeping work (like Error structures, etc), which will be done once the API is more or less finalized.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+        self.spawn_workers();
+
+        // Get already existing cache
+        let mut cache = match AddressCache::from_file(&self.cache_filename) {
```
It looks like the cache is loaded in the constructor as well, maybe you can just store it in the struct to avoid re-loading it here

rajarshimaitra replied:
Hmm.. I was trying to reduce the memory load of the structure. `cache` is initiated at the constructor and `fetch()`.

In the constructor, it's created to feed the seeds to the workers.

In `fetch`, it's created to update the disk storage.

In my head, I think these will be called very few times.

constructor: Once
fetch: Only when available fresh address reserve depletes under a threshold.

For this reason, I figured it might not be clever to carry a large cache around with the address manager. And only create when needed.

rajarshimaitra replied:
Also note that the cache is essentially duplicated inside `Directory` which is part of the address manager. So thought it probably doesn't make sense to carry around duplicate data, only to use them occasionally.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+    /// - check if address is in banlist
+    /// - run crawlers until buffer requirement is matched
+    /// - flush the current cache into disk
+    pub fn fetch(&mut self) {
```
Ideally I would change this to return a `Result` and remove the unwraps inside

rajarshimaitra replied:
Ideally, all `pub` functions should give back `Result`. Working on it.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+        }
+
+        // When completed, flush the cache
+        cache.write_to_file(&self.cache_filename).unwrap();
```
Maybe it makes sense to write the cache periodically and not just at the end?

rajarshimaitra replied:
Yes it does. But that involves threading. I thought of that first in general as it made more sense, but having multiple threads made stuff difficult. I already went into deadlock scenarios with just worker threads. So decided to keep stuff simple and ended up writing most of the things single-threaded and optimized around executions.  

As mentioned in https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660353392 we can expect `fetch` to be called occasionally. As a result, cache will also update occasionally, and we will know exactly when it updates, only in a `fetch` call. 

Thus it made sense to not give its own dedicated loop and just use file i/o only when needed.

johncantrell97 replied:
At one point I had it writing the cache to file inside of the fetch() message loop from workers anytime I updated the memory cache.  I ran into a 'too many open files' issue though and so moved it out.  Could probably get away with some kind or probabilistic flush if we want to update it more often but not every iteration of the loop.

rajarshimaitra replied:
Yes, That's the reason I took it out of the loop. So currently it's not flushing at each iteration. It flushes after the loop completes. So that problem doesn't exist now.

https://github.com/bitcoindevkit/bdk/blob/23fe55127e0a4c6871df0ff66ce8897aeff78143/src/blockchain/compact_filters/address_manager.rs#L541-L542

Also, note that the loop will not start in case the buffer requirement is already fulfilled. Flush will happen only once per `fetch()` or `ban()` call. Because these are the only calls where we expect the cache to be updated.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+    pub fn ban_peer(&mut self, addrs: &SocketAddr, cbf: bool) {
+        let mut cache = AddressCache::from_file(&self.cache_filename).unwrap();
+
+        cache.add_to_banlist(*addrs, cbf);
```
I think this is not being written back to disk. Adding the cache as a struct member would allow to reuse it here as well

rajarshimaitra replied:
Ah right, it wasn't being written to disk. Fixed that. 

Because of the reasons explained here https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660353392 I am inclining against adding the cache to the structure.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+                    );
+
+                    if let Ok(peer) = potential_peer {
+                        peer.send(NetworkMessage::GetAddr).unwrap();
```
I would remove the unwrap here and right down below, the `send` can fail if the underlying socket disconnects

rajarshimaitra replied:
Agreed, will be fixed in error handling commits.

afilini commented on src/blockchain/compact_filters/p2p.rs:
```diff
@@ -0,0 +1,646 @@
+                        peer.close().unwrap_or(());
+                    }
+                }
+                None => continue,
```
Maybe if we can't connect we should "ban" the peer?

I wonder if it makes sense to add a "ban until" timestamp to the cache, so that we can differentiate between a permanent ban (peer misbehaving) or a temporary "i can't connect right now" ban

rajarshimaitra replied:
Hmm, probably. Currently what it does is if it can't connect it will simply skip. I am not sure if connection failure should result into ban, as nodes can be offline temporarily, they may connect in next try. 

I think we definitely need to define banning more granularly, and I think best place to do that will be the `peermanager` (which I am sketching now, and it will use address manager inside). I have put a ban score flag into PeerManager, so we can have more granular banning logic like core does. 

And the objective here is the cache will maintain only the final ban list such that once it's banned it will never get into discovery again.

And then in peer manager, we can design different levels of banning, and decide whether to tell the address manager to "cache ban" the address or not.

rajarshimaitra commented:
Thanks @afilini for the review. Tried to address all of them. Overall the intention was to not involve threads and try to optimize around function calls and executions. That guided lot of the decisions. I think I have got it into a satisfactory level of efficiency that with an existing cache an address DB can be populated in few seconds with freshly connected cbf and non_cbf nodes. 

There's probably more optimizations possible.   

Following are the next TODOs after which I think I can get it out of draft.

- Handle results and options everywhere. This probably will also involve changes in `Peer` module as I think it should have its dedicated errors, and that decision will affect address manager errors too.
- Few minor possible changes in the `AddressManager` API as I sketch the basics of `PeerManager`.
- Few more tests (looking for ideas).

rajarshimaitra commented:
Updated with Error handling. 

A new `PeerError` structure added.

All the other internal module-level errors will be finally combined into `CompactFiltersError`.

rajarshimaitra committed:
Add PeerError structure in peer module

This adds a new PeerError structure in the peer module. To handle all
the peer related errors. PeerErrors contains all the mempool errors too,
for now. Later if we have a more complex mempool, we might decide to
have its own dedicated error.

PeerError is to be included in the global CompactFiltersError type.

rajarshimaitra committed:
Add Address Manager

Add the final address manager structures. Address Manager is responsible
for finding new peers from the network. It maintains a directory with
cbf/non_cbf catagories of peer. A Peer Manager will query this
Address Manager to provide it with fresh addresses.

Currently Address Manager is single threaded. When a user calls it to
fetch the network, it will block the call until fetch completes.

It knows when to stop fetching via config variable MIN_CBF_BUFFER and
MIN_NONCBF_BUFFER. And it stops fetching when buffer requirement is met.

Co-authored-by: John Cantrell <johncantrell97@protonmail.com>

rajarshimaitra committed:
Update compact_filters module

Compact filters module is updated to include the new Address Manager.

The new PeerError structure is added into CompactFiltersError to have
compatibility with existing APIs.

Minor cargo fmt nit fixes

rajarshimaitra commented:
Updated the commit history consolidating all the changes.

Address Manager seems ready from my side. All the APIs I think are there required to make a basic Peer Manager.

This is now ready for public review.

rajarshimaitra committed:
Return Option when getting addresses

When getting addresses return Option with Some vectors, or None in
case of empty vectors.

This makes handling the case of getting empty result easier.

rajarshimaitra commented:
Rebased

notmandatory commented:
I had to cancel the CI workflows for this branch because the `compact_filters` tests seems to be hanging. Could this be some thread locking issue? @rajarshimaitra have you seen this happen on your local testing?

rajarshimaitra committed:
Update address manager test

The two tests have been combined into a single one. Added extra
log points for better debugging in CI if it gets stuck.

rajarshimaitra commented:
Hmm.. It seems the test got stuck somewhere. Which is not happening in my local.

I have combined the two tests into a single one, running two tests parallelly can provide weird errors. 
Added extra log points to see what step got stuck if it happens again.

There was a hang issue as I mentioned in https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660344444, which after some refactoring didn't occur anymore in my local. It could be this, but unlikely. Maybe others can confirm from their locals. If it happens then one will see the pending number goes to zero in the test log, and the test hangs.

It seems to be working fine in my local and here's what I see.

```
> Executing task: cargo test --package bdk --lib -- blockchain::compact_filters::address_manager::test --nocapture <

   Compiling bdk v0.8.1-dev (/home/raj/github-repo/bdk)
    Finished test [unoptimized + debuginfo] target(s) in 9.25s
     Running target/debug/deps/bdk-fdc911b1c7b1070c

running 1 test
Starting manager and initial fetch
P2P Discovery: 4382 queued, 688 visited, 221 connected, 5 cbf_enabled
Completed Initial fetch
test blockchain::compact_filters::address_manager::test::test_address_manager ... test blockchain::compact_filters::address_manager::test::test_address_manager has been running for over 60 seconds
Starting new fetch with previous cache
P2P Discovery: 1359 queued, 40 visited, 20 connected, 5 cbf_enabled
Completed new fetch()
Time taken for initial crawl: 44.709318174s
Time taken for next crawl 2.871860807s
Checking buffer management
P2P Discovery: 9808 queued, 2916 visited, 465 connected, 10 cbf_enabled
Buffer management passed
test blockchain::compact_filters::address_manager::test::test_address_manager ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 203 filtered out; finished in 157.53s  
```

The test takes 157.53 secs. Which seems much lower than github action would mind. But it might take a bit more in the CI server. If that's the issue, then we should ignore this test for now until we figure out how to make it faster.

rajarshimaitra commented:
I am not sure why this test https://github.com/bitcoindevkit/bdk/pull/378/checks?check_run_id=3034931064 is stuck while all others are completed.

I am out of suggestions here. If the high test time was a problem, github would have complained in other tests too. Not really sure whats the problem. It seems it got cancelled before the address manager could initialise. The test will try to create a cache file in run directory also, but I suppose that should not be an issue.

Probably we should just ignore this test for now?

notmandatory commented:
I had to cancel the "Code Coverage" workflow, it was taking +2hrs for some reason. It seemed to be stalled on this: 
```
test blockchain::compact_filters::address_manager::test::test_address_manager has been running for over 60 seconds
```

I think it's OK to disable this test for now and then we can see if we can construct some focused regtest tests with https://github.com/RCasatta/bitcoind.

rajarshimaitra commented:
@notmandatory noted. 

We can disable the test and merge, or we can fix this here itself. 

I am working on creating a controlled regtest to do the test. Maybe in few more days. I think its better to get this thing done now, than come back later. 

We can merge this after that.

notmandatory commented:
> @notmandatory noted.
> 
> We can disable the test and merge, or we can fix this here itself.

After the long discussion on Discord I agree best to disable the `test_address_manager` test for now so we can merge this and have all the code together in the `compact_filters` branch.

notmandatory reviewed (approved):
ACK, merging to `compact_filters` branch.

notmandatory committed:
Ignore wip  test

notmandatory commented:
Looks like we lost the `#[ignore]` on the `test_address_manager()` test. I pushed it back up and if all the rest of the tests pass I should be able to do the merge.
